### DISCUSSION TITLE
How to get primary id from within a transaction

### DISCUSSION THREAD

**Discussion by jeffa5** [microsoft/CCF#4337] [Q&A]
When executing a transaction (or on a command endpoint), it would be useful to be able to get the id of the leader (primary), similar to https://microsoft.github.io/CCF/main/operations/operator_rpc_api.html#get--node-consensus provides.

I can't see a way to get it from the [internal map](https://microsoft.github.io/CCF/main/audit/builtin_maps.html#nodes-info) either and the logic the above endpoint uses isn't exposed.

**Comment by eddyashton** [comment]
The reason this isn't exposed is that the primary changes _outside_ of the transaction order. So different nodes may return different primaries for otherwise identical requests. We're wary of letting apps make decisions within a transaction that depend on their (transient, local) opinion of the current primary.

What's the use case?

**Reply by jeffa5** [reply]
Building a custom status endpoint in the application that can return cluster-level information in a custom format. In this instance gRPC. It being only on readonly and command endpoints would be all I need. I agree that it could be troublesome if users included it in transactions that wrote to the maps. It isn't expected to be 100% accurate either, just a best guess so staleness is ok.
